datastore/providers/opensearch_datastore.py

```diff
@@ -0,0 +1,422 @@
+                print(f"Deleted document Response from delete by document_id : {response}")
+            except Exception as e:
+                print(f"Error deleting ids: {ids} with exception : {e}")
+                raise e
+
+        return True
+
+    async def __setup_and_validate_data_store(self):
+        await self.__check_knn_plugin_present()
+        await self.__validate_index_setup()
+        await self.__validate_cluster_health()
+
+    async def __check_knn_plugin_present(self):
+        response = str(await self.async_client.cat.plugins())
+        if response.__contains__('opensearch-knn') is False:
+            raise Exception("The OpenSearch Data store doesn't contains the K-NN plugin. Please install K-NN plugin. "
+                            "Ref: https://opensearch.org/docs/latest/install-and-configure/plugins/ ")
+        print("OpenSearch K-NN plugin present in OpenSearch Cluster")
+
+    async def __validate_index_setup(self):
+        if (await self.async_client.indices.exists(OPENSEARCH_INDEX_NAME)) is True:
+            print(f"Index with name : {OPENSEARCH_INDEX_NAME}, already created.")
+            await self.__check_if_vector_field_present()
+        else:
+            await self.__create_index()
+
+    async def __check_if_vector_field_present(self):
+        mappings_response = str(await self.async_client.indices.get_mapping(index=OPENSEARCH_INDEX_NAME))
+        if mappings_response.__contains__("knn_vector") is False:
+            raise Exception("There is no field containing knn_vector as field type. Please make sure to create a k-nn "
+                            "vector field in the index mappings. Reference: "
+                            "https://opensearch.org/docs/latest/search-plugins/knn/knn-index/")
+
+    async def __create_index(self):
+        print(f"Creating index with name {OPENSEARCH_INDEX_NAME}")
+        response = await self.async_client.indices.create(OPENSEARCH_INDEX_NAME, body=index_body)
+
+        if response['acknowledged'] is True:
+            print("Index {} Created successfully.".format(OPENSEARCH_INDEX_NAME))
+        else:
+            raise Exception(f"Unable to create Index with name: {OPENSEARCH_INDEX_NAME}, response: {response}")
+
+    async def __validate_cluster_health(self):
+        cluster_health = await self.async_client.cluster.health()
+        if cluster_health["status"] == "green":
+            print("Cluster Status is green.")
+        else:
+            print(f"Cluster health is : {cluster_health['status']}, which can lead to failures during upsert, query "
+                  f"and delete operations.")
+
+    @staticmethod
+    def __get_opensearch_metadata(metadata: Optional[DocumentChunkMetadata] = None) -> Dict[str, Any]:
+        if metadata is None:
+            return {}
+        opensearch_metadata = {}
+
+        for field, value in metadata.dict().items():
+            if type(value) is Source:
```

Comment:
@navneet1v without this condition, we get error:
```
Error: Cannot parse argument of type None.
```

```suggestion
            if type(value) is Source:
                if type(value) is Source:
                    ....
```

Reply:
sure let me fix this.

Reply:
@Venkat2811 I tried with this input and it worked for me.

```
{
  "documents": [
    {
      "id": "doc1",
      "text": "This is my string",
      "metadata": {
        "source_id": "string",
        "url": "string",
        "created_at": "2017/10/10",
        "author": "string"
      }
    }
  ]
}
```

and 

```
{
  "documents": [
    {
      "id": "doc2",
      "text": "This is my string",
      "metadata": {
        "source": "email",
        "url": "string",
        "created_at": "2017/10/10",
        "author": "string"
      }
    }
  ]
}
```

both of them worked for me. Not sure when there will be a case that will have the key but the value is None. I can add an explicit check, but can you tell me for what input failed for you?

Reply:
After deleting `created_at` and even whole `metadata` from default input `http://localhost:8000/docs#/default/upsert_upsert_post`

```
{
  "documents": [
    {
      "id": "string",
      "text": "string",
      "metadata": {
        "source": "email",
        "source_id": "string",
        "url": "string",
        "author": "string"
      }
    }
  ]
}
```

and

```
{
  "documents": [
    {
      "id": "string",
      "text": "string"
    }
  ]
}
```

this input works with other providers btw.

Reply:
Thanks for providing the input. I have fixed the issue, using the same code which you suggested. I didn't expect key to be present and the value to be None for the key which are not provided by user in here. and the issue is only happening for created_at filed only because its a date and we are parsing it.
```
for field, value in metadata.dict().items():
```